### PR TITLE
Send Team Emails

### PR DESCRIPTION
**Pull request by k3ntako**
Allows any user to send an email to the team.
- It's assumed that you are emailing every member unless they are exclude.
- Can add non-team members.

MSK Admin is part of our team. What should we do with that account?

### CODE REVIEW

File: client/pages/Email/ExcludeForm.js [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+
+    return <Card className={styles.card}>
+      <div className={rStyles.formTitle}>Exclude a Member from Email</div>
+      <p>
+        Email will be sent to all coaches and members, however, you can choose to exclude any recipients.
+        <br /><br />
+        You can include a member that was previously excluded by selecting them in the same text box below.
+      </p>
+
+      <Input
```
Comment (mienaikoe) [inline client/pages/Email/ExcludeForm.js]: Looks like this is an autocomplete. Any thoughts on using a Table instead? I see some pros and cons to each, but thought I'd open up a discussion about it.

File: public/index.css [hunk 2]
```diff
@@ -5,6 +5,7 @@
 
 html{
   font-size: 12px;
+  height: 100%;
```
Comment (mienaikoe) [inline public/index.css]: No! This constrains the document to be the viewport height. Scrolling gets all out of wack. Why is this here?

File: server/models/Email.js [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+var EmailSchema = new Schema({
+  _id: ObjectId,
+  sender: { type: ObjectId, ref: 'Person', required: true },
+  subject: { type: String, trim: true },
+  team: { type: ObjectId, ref: 'Team', required: true, index: true },
+  exclude: [
+    { type: ObjectId, ref: 'Person' }
+  ],
+  nonMember: [
+    { type: Object }
```
Comment (mienaikoe) [inline server/models/Email.js]: Let's make a schema for nonMembers instead of a generic object

File: server/models/Email.js [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+  ],
+  emailBody: {
+    type: Object,
+    required: true,
+    default: draftRaw
+  },
+  sendTime: { type: Date, default: null},
+  rejected: [
+    { type: String }
```
Comment (mienaikoe) [inline server/models/Email.js]: Would boolean suffice here?

File: server/models/Email.js (hunk 3 above)
Comment (k3ntako) [reply server/models/Email.js]: so, this is a list of emails that were "rejected" (I think that means bounced back)

File: server/models/Email.js (hunk 4 above)
Comment (k3ntako) [reply server/models/Email.js]: this is a list of emails that were "rejected" (I think that means bounced back)

Comment (mienaikoe) [reply server/models/Email.js]: Ahh ok. so then the purpose is to tell coach which email didn't get received. Aight that's fine.

File: server/tests/TestHelper.js [hunk 5]
```diff
@@ -68,6 +71,31 @@ module.exports = {
...
+        body: {
+          first_name: 'Team',
+          last_name: 'Mate',
+          email: email || `TeamMate${randomnum()}@wakeBird.com`,
+          password: "testtesttest",
+        },
+      }
+      const res = {
+        json: function(response){
+          test_teammate = response;
```
Comment (mienaikoe) [inline server/tests/TestHelper.js]: I want to avoid any more saving of variables on this file. The variables should be saved in their respective tests, because the order and sequence of the tests can't be guaranteed, and the tests shouldn't rely on each other to have already called certain functions.

This isn't documented anywhere, so I don't blame you. We can talk about it a bit more on Monday. but for now, please remove test_teammate from this file, and just use the response from this function.

File: server/utilities/team_email.js [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+
+  return accumulator;
+};
+
+const recipientList = (teamEmail) => {
+  const members = teamEmail.team.coaches.concat(teamEmail.team.members).concat(teamEmail.nonMember);
+  const exclude = teamEmail.exclude.map(person => person.email);
+
+  //remove duplicate emails and exclude members
+  const recipients = members.reduce((accumulator, currentValue) => {
```
Comment (mienaikoe) [inline server/utilities/team_email.js]: This isn't the optimum way to create the recipients list. This is the sort of thing that will be in an interview. Your challenge is to find a quicker way to do this.

File: client/pages/Team/EmailDraftTable.js [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+      return <Error
+        errorTitle={"No Drafts!"}
+        message={composeButton}
+        icon="inbox"
+        />
+    }
+
+    return <>
+      <div className={styles.draftHeading}>
+        <h2>Drafts</h2>
```
Comment (mienaikoe) [inline client/pages/Team/EmailDraftTable.js]: I recommend using h3 instead of h2

File: client/pages/Team/EmailTableTemplate.js [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+    tableHead = <TableHead>
+      <TableCell>Subject</TableCell>
+      <TableCell>Sent</TableCell>
+    </TableHead>
+  }else{
+    tableHead = <TableHead>
+      <TableCell className={styles.checkBox}>
+        <Checkbox
+          checked={props.checkAll}
+          onChange={() => props.handleChange('checkAll')} />
```
Comment (mienaikoe) [inline client/pages/Team/EmailTableTemplate.js]: This didn't seem to work when I tested it locally. It just stayed unchecked when I clicked it on drafts.

File: client/pages/Team/EmailDraftTable.js [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+        <div className={styles.buttons}>
+          <div>
+            <Button
+              raised primary
```
Comment (mienaikoe) [inline client/pages/Team/EmailDraftTable.js]: Doesn't need to be raised

File: client/pages/Team/EmailDraftTable.js [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+        errorTitle={"You're Offline!"}
+        message={"Check back when you're back online"}
+        icon="offline_bolt"
+        />
+    }
+
+    const composeButton = <Button
+      raised primary
+      label="Compose"
+      onClick={this.compose.bind(this)} />
```
Comment (mienaikoe) [inline client/pages/Team/EmailDraftTable.js]: you could make this an icon+text button to give it some visual hook. Same with the delete button.

File: client/pages/Email/index.js [hunk 11]
```diff
@@ -0,0 +1,184 @@
...
+          index={tabIndex}
+          onChange={this.handleTabChange.bind(this)}
+          theme={{
+            tab: styles.tab,
+            tabs: styles.tabs
+          }}>
+          <Tab label='Email' className={styles.tabHeader}>
+            <div className={styles.spaceBetweenButton}>
+              <div className={styles.link}>
+                <Link to="/#email">Back to Inbox</Link>
```
Comment (mienaikoe) [inline client/pages/Email/index.js]: 1) It's not an inbox if the list is all outgoing mail
2) This is the sort of thing crumbs in the header are for. If you think that it takes up so much space that this is the only place in the whole site that the header shouldn't be, then perhaps we should shrink the header a bit. Let's talk about this Monday.

File: client/pages/Email/index.js [hunk 12]
```diff
@@ -0,0 +1,184 @@
...
+              subjectChangeHandler={this.subjectChangeHandler.bind(this)}
+              isSending={isSending}
+              setIsSending={this.setVerb.bind(this, "isSending")}
+              isSaved={isSaved}
+              setIsSaved={this.setIsSaved.bind(this)}
+              isSaving={isSaving}
+              setIsSaving={this.setVerb.bind(this, "isSaving")}
+              /> }
+          </Tab>
+          <Tab label='Recipients' className={styles.tabHeader}>
```
Comment (mienaikoe) [inline client/pages/Email/index.js]: I completely missed the recipients tab the first couple times I was using this. Maybe it shouldn't be in tabs. I get that it's a relatively large set of options, but perhaps most of them can be put into modals instead of cards that fill up the page space.

File: client/pages/Email/ExcludeForm.js (hunk 1 above)
Comment (mienaikoe) [reply client/pages/Email/ExcludeForm.js]: Also, I have a more general case of a person autocomplete coming up in the assignments branch that you can use instead of this if you still think autocomplete is the right way to go.

File: client/pages/Email/ExcludeForm.js [hunk 13]
```diff
@@ -0,0 +1,167 @@
...
+  include(){
+    this.props.email.excludeMember(this.state.selectedPerson.id, false).then(() => {
+      this.setState({search: "", selectedPerson: null}, this.props.forceUpdate)
+    });
+  }
+
+  render(){
+    const { isOnline, isServerConnected } = this.props;
+
+    const disableButtons = !selectedPerson || !isOnline || isServerConnected;
```
Comment (mienaikoe) [inline client/pages/Email/ExcludeForm.js]: selectedPerson isn't defined until line 104. This makes `disableButtons` disabled forever

File: client/pages/Email/NonMemberForm.js [hunk 14]
```diff
@@ -0,0 +1,142 @@
...
+          onClick={this.add.bind(this)}
+          disabled={disableButtons || !this.valid()}/>
+      </div>
+
+      <div className={rStyles.formTitle} id={rStyles.removeNonMember}>
+        Remove Non-Member from Email
+      </div>
+      <p>Remove recipients that were added above.</p>
+      <div className={rStyles.inputRow}>
+        <Input
```
Comment (mienaikoe) [inline client/pages/Email/NonMemberForm.js]: this should be done via a button. There's no need to re-enter information that the app already has.

File: public/index.css (hunk 2 above)
Comment (k3ntako) [reply public/index.css]: Put buttons at bottom and make Draft.js the height of content.

Comment (k3ntako) [reply public/index.css]: make delete button an icon

File: server/utilities/team_email.js (hunk 6 above)
Comment (k3ntako) [reply server/utilities/team_email.js]: don't need to exclude `nonMembers`

use filter for exclude and take it out of reducer

File: client/pages/Email/index.js (hunk 12 above)
Comment (k3ntako) [reply client/pages/Email/index.js]: Add member modal

exclusion inclusion and deletion of non member can be done on list

list can be under editor

File: server/utilities/team_email.js [hunk 15]
```diff
@@ -8,24 +8,11 @@ const convertFromRaw = require('draft-js').convertFromRaw;
...
-  if(!accumulator.includes(address) && !exclude.includes(address)){
-    accumulator.push(address)
-  }
-
-  return accumulator;
-};
-
 const recipientList = (teamEmail) => {
-  const members = teamEmail.team.coaches.concat(teamEmail.team.members).concat(teamEmail.nonMember);
+  const members = teamEmail.team.members.map(person => person.email);
```
Comment (k3ntako) [inline server/utilities/team_email.js]: I wasn't too sure about this, but coaches are members, right?

Comment (mienaikoe) [reply server/utilities/team_email.js]: In the migrations I specified that coaches were also members. So this should be ok.
